### PR TITLE
Addition of JS API

### PR DESCRIPTION
**Pull request by Secretmapper**
+ Change queries without reloading the page   `iframe.contentWindow.postMessage({cmd:'run_query', value:['query']}, "*");`
+ Cachable queries (can be run only)
`iframe.contentWindow.postMessage({cmd:'run_query', value:['query', true]}, "*");`
+ Catch overpass errors/alerts
Listen to "message" event

NOTE: Everything must be JSON parsed/stringify

### CODE REVIEW

File: js/map.js [hunk 1]
```diff
@@ -3,6 +3,36 @@ function htmlentities(str) {
...
+          overpass.handlers["onQueryError"] = function(errmsg) {
+              parent.postMessage(JSON.stringify({handler: "onQueryError",
+                                                 msg: errmsg }), "*");
+          };
+          break;
+        case 'update_map':
+          settings.code["overpass"] = data.value[0];
+          ide.update_map();
+          break;
+        case 'run_query':
```
Comment (tyrasd) [inline js/map.js]: this could use a better name (to make sure it's immediately associated as the caching endpoint)

File: js/map.js [hunk 2]
```diff
@@ -3,6 +3,36 @@ function htmlentities(str) {
...
 }
 $(document).ready(function() {
+  // main map cache
+  var cache = {};
+
+    window.addEventListener("message", function (evt) {
+      var data = JSON.parse(evt.data);
+      switch(data.cmd){
+        case 'catch_alert':
+          overpass.handlers["onAjaxError"] = function(errmsg) {
```
Comment (tyrasd) [inline js/map.js]: We should probably store any previously set event handlers and call them inside the function?

Comment (Secretmapper) [inline js/map.js]: AjaxErrors/QueryErrors only triggers an alert, which is synchronous and blocking - so calling it would be problematic.

Comment (tyrasd) [inline js/map.js]: true. forget what I said

**Comment by tyrasd** [comment]
this should now be live on http://overpass-turbo.eu/master/

**Pull request merged**